// @angular-architects/module-federation-tools/bundles/angular-architects-module-federation-tools.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('@angular-architects/module-federation'), require('@angular/router'), require('@angular/platform-browser')) :
    typeof define === 'function' && define.amd ? define('@angular-architects/module-federation-tools', ['exports', '@angular/core', '@angular/common', '@angular-architects/module-federation', '@angular/router', '@angular/platform-browser'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global['angular-architects'] = global['angular-architects'] || {}, global['angular-architects']['module-federation-tools'] = {}), global.ng.core, global.ng.common, global.moduleFederation, global.ng.router, global.ng.platformBrowser));
}(this, (function (exports, i0, common, moduleFederation, i1, platformBrowser) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () {
                            return e[k];
                        }
                    });
                }
            });
        }
        n['default'] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from) {
        for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
            to[j] = from[i];
        return to;
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    // eslint-disable-next-line @angular-eslint/component-class-suffix
    var WebComponentWrapper = /** @class */ (function () {
        function WebComponentWrapper(route) {
            this.route = route;
        }
        WebComponentWrapper.prototype.ngOnChanges = function () {
            if (!this.element)
                return;
            this.populateProps();
        };
        WebComponentWrapper.prototype.populateProps = function () {
            for (var prop in this.props) {
                this.element[prop] = this.props[prop];
            }
        };
        WebComponentWrapper.prototype.setupEvents = function () {
            for (var event in this.events) {
                this.element.addEventListener(event, this.events[event]);
            }
        };
        WebComponentWrapper.prototype.ngAfterContentInit = function () {
            var _a;
            return __awaiter(this, void 0, void 0, function () {
                var options, error_1;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            options = (_a = this.options) !== null && _a !== void 0 ? _a : this.route.snapshot.data;
                            _b.label = 1;
                        case 1:
                            _b.trys.push([1, 3, , 4]);
                            return [4 /*yield*/, moduleFederation.loadRemoteModule(options)];
                        case 2:
                            _b.sent();
                            this.element = document.createElement(options.elementName);
                            this.populateProps();
                            this.setupEvents();
                            this.vc.nativeElement.appendChild(this.element);
                            return [3 /*break*/, 4];
                        case 3:
                            error_1 = _b.sent();
                            console.error(error_1);
                            return [3 /*break*/, 4];
                        case 4: return [2 /*return*/];
                    }
                });
            });
        };
        return WebComponentWrapper;
    }());
    WebComponentWrapper.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: WebComponentWrapper, deps: [{ token: i1__namespace.ActivatedRoute }], target: i0__namespace.ɵɵFactoryTarget.Component });
    WebComponentWrapper.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.0.0", type: WebComponentWrapper, selector: "mft-wc-wrapper", inputs: { options: "options", props: "props", events: "events" }, viewQueries: [{ propertyName: "vc", first: true, predicate: ["vc"], descendants: true, read: i0.ElementRef, static: true }], usesOnChanges: true, ngImport: i0__namespace, template: '<div #vc></div>', isInline: true });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: WebComponentWrapper, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'mft-wc-wrapper',
                        template: '<div #vc></div>',
                    }]
            }], ctorParameters: function () { return [{ type: i1__namespace.ActivatedRoute }]; }, propDecorators: { vc: [{
                    type: i0.ViewChild,
                    args: ['vc', { read: i0.ElementRef, static: true }]
                }], options: [{
                    type: i0.Input
                }], props: [{
                    type: i0.Input
                }], events: [{
                    type: i0.Input
                }] } });

    var ModuleFederationToolsModule = /** @class */ (function () {
        function ModuleFederationToolsModule() {
        }
        return ModuleFederationToolsModule;
    }());
    ModuleFederationToolsModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: ModuleFederationToolsModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    ModuleFederationToolsModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: ModuleFederationToolsModule, declarations: [WebComponentWrapper], imports: [common.CommonModule], exports: [WebComponentWrapper] });
    ModuleFederationToolsModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: ModuleFederationToolsModule, imports: [[
                common.CommonModule
            ]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0__namespace, type: ModuleFederationToolsModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [
                            common.CommonModule
                        ],
                        declarations: [
                            WebComponentWrapper
                        ],
                        exports: [
                            WebComponentWrapper
                        ]
                    }]
            }] });

    var packageNamespace = '@angular-architects/module-federation-tools';
    function getGlobalState() {
        var globalState = window;
        globalState[packageNamespace] = globalState[packageNamespace] || {};
        return globalState[packageNamespace];
    }
    function getGlobalStateSlice(selector) {
        var globalState = getGlobalState();
        return selector ? selector(globalState) : globalState;
    }
    function setGlobalStateSlice(slice) {
        return Object.assign(getGlobalState(), slice);
    }

    function startsWith(prefix) {
        return function (url) {
            var fullUrl = url.map(function (u) { return u.path; }).join('/');
            if (fullUrl.startsWith(prefix)) {
                return ({ consumed: url });
            }
            return null;
        };
    }
    function endsWith(prefix) {
        return function (url) {
            var fullUrl = url.map(function (u) { return u.path; }).join('/');
            if (fullUrl.endsWith(prefix)) {
                return ({ consumed: url });
            }
            return null;
        };
    }
    function connectRouter(router, useHash) {
        if (useHash === void 0) { useHash = false; }
        var url;
        if (!useHash) {
            url = "" + location.pathname.substr(1) + location.search;
            router.navigateByUrl(url);
            window.addEventListener('popstate', function () {
                router.navigateByUrl(url);
            });
        }
        else {
            url = "" + location.hash.substr(1) + location.search;
            router.navigateByUrl(url);
            window.addEventListener('hashchange', function () {
                router.navigateByUrl(url);
            });
        }
    }

    var ngZoneSharing = true;
    var platformSharing = true;
    var legacyMode = true;
    function getMajor(version) {
        var pre = version.match(/\d+/)[0];
        var post = version.match(/-.*/);
        if (!pre) {
            throw new Error('Cound not identify major version: ' + version);
        }
        if (post) {
            return pre + post[0];
        }
        return pre;
    }
    function getLegacyPlatformCache() {
        var platformCache = window;
        platformCache.platform = platformCache.platform || {};
        return platformCache;
    }
    function getLegacyPlatform(key) {
        var platform = getLegacyPlatformCache().platform[key];
        /**
         * If dependencies are not shared, platform with same version is different
         * and shared platform will not be returned.
         */
        return platform instanceof i0.PlatformRef ? platform : null;
    }
    function setLegacyPlatform(key, platform) {
        getLegacyPlatformCache().platform[key] = platform;
    }
    function getLegacyNgZone() {
        return window['ngZone'];
    }
    function setLegacyNgZone(zone) {
        window['ngZone'] = zone;
    }
    /**
     * LEGACY IMPLEMENTATIONS END
     */
    function getPlatformCache() {
        return getGlobalStateSlice(function (state) { return state.platformCache; }) || setGlobalStateSlice({
            platformCache: new Map()
        }).platformCache;
    }
    function setPlatform(version, platform) {
        if (platformSharing) {
            legacyMode && setLegacyPlatform(version.full, platform);
            getPlatformCache().set(version, platform);
        }
    }
    function getPlatform(options) {
        if (!platformSharing) {
            return options.platformFactory();
        }
        var versionResult = options.version();
        var version = versionResult === i0.VERSION.full ? i0.VERSION : versionResult;
        var versionKey = typeof version === 'string' ? version : version.full;
        var platform = getPlatformCache().get(version) ||
            (legacyMode && getLegacyPlatform(versionKey));
        if (!platform) {
            platform = options.platformFactory();
            setPlatform(i0.VERSION, platform);
            options.production && i0.enableProdMode();
        }
        return platform;
    }
    function getNgZone() {
        return getGlobalStateSlice(function (state) { return state.ngZone; }) || getLegacyNgZone();
    }
    function shareNgZone(zone) {
        if (ngZoneSharing) {
            legacyMode && setLegacyNgZone(zone);
            setGlobalStateSlice({ ngZone: zone });
        }
    }
    function bootstrap(module, options) {
        var _a;
        ngZoneSharing = options.ngZoneSharing !== false;
        platformSharing = options.platformSharing !== false;
        legacyMode = options.activeLegacyMode !== false;
        options.platformFactory = options.platformFactory || (function () { return platformBrowser.platformBrowser(); });
        options.version = options.version || (function () { return i0.VERSION; });
        if (ngZoneSharing && !((_a = options.compilerOptions) === null || _a === void 0 ? void 0 : _a.ngZone)) {
            options.compilerOptions = options.compilerOptions || {};
            options.compilerOptions.ngZone = getNgZone();
        }
        return getPlatform(options).bootstrapModule(module, options.compilerOptions).then(function (ref) {
            if (options.appType === 'shell') {
                shareShellZone(ref.injector);
            }
            else if (options.appType === 'microfrontend') {
                connectMicroFrontendRouter(ref.injector);
            }
            return ref;
        });
    }
    function shareShellZone(injector) {
        var ngZone = injector.get(i0.NgZone, null);
        if (!ngZone) {
            console.warn('No NgZone to share found');
            return;
        }
        shareNgZone(ngZone);
    }
    function connectMicroFrontendRouter(injector) {
        var router = injector.get(i1.Router);
        if (!router) {
            console.warn('No router to connect found');
            return;
        }
        connectRouter(router);
    }

    /**
     * Generated bundle index. Do not edit.
     */

    exports.ModuleFederationToolsModule = ModuleFederationToolsModule;
    exports.WebComponentWrapper = WebComponentWrapper;
    exports.bootstrap = bootstrap;
    exports.connectRouter = connectRouter;
    exports.endsWith = endsWith;
    exports.getMajor = getMajor;
    exports.shareNgZone = shareNgZone;
    exports.startsWith = startsWith;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
