// @angular-architects/module-federation-tools/esm2015/lib/mf-tools.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { WebComponentWrapper } from './web-components/web-component-wrapper';
import * as i0 from "@angular/core";
export class ModuleFederationToolsModule {
}
ModuleFederationToolsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0, type: ModuleFederationToolsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ModuleFederationToolsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0, type: ModuleFederationToolsModule, declarations: [WebComponentWrapper], imports: [CommonModule], exports: [WebComponentWrapper] });
ModuleFederationToolsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0, type: ModuleFederationToolsModule, imports: [[
            CommonModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.0", ngImport: i0, type: ModuleFederationToolsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [
                        WebComponentWrapper
                    ],
                    exports: [
                        WebComponentWrapper
                    ]
                }]
        }] });
